Add `no-default-export` + docs/tests

isiahmeadows opened the pull request:
Fixes #889

isiahmeadows commented:
Fixes #889

isiahmeadows commented:
The failing tests appear to be irrelevant (and the same error that appears in `master`).

j-f1 commented on tests/src/rules/no-default-export.js:
```diff
@@ -0,0 +1,123 @@
+      parser: 'babel-eslint',
+      }),
+
+    // ...SYNTAX_CASES,
```
What does this mean?

isiahmeadows replied:
It was more or less adapted from [`prefer-default-export`'s tests](https://github.com/benmosher/eslint-plugin-import/blob/master/tests/src/rules/prefer-default-export.js#L84).

j-f1 commented on README.md:
```diff
@@ -91,6 +91,7 @@ This plugin intends to support linting of ES2015+ (ES6+) import/export syntax, a
 [`no-unassigned-import`]: ./docs/rules/no-unassigned-import.md
 [`no-named-default`]: ./docs/rules/no-named-default.md
 [`no-anonymous-default-export`]: ./docs/rules/no-anonymous-default-export.md
+[`no-default-export`]: ./docs/rules/no-default-export.md
```
You should mention this under one of the headers so it is visible in the README.

isiahmeadows replied:
Good catch! I completely missed that.

isiahmeadows replied:
Fixed!

j-f1 commented on src/rules/no-default-export.js:
```diff
@@ -0,0 +1,40 @@
+              node.exported.name === 'default') {
+            context.report({
+              node: ast,
+              message: 'Prefer named exports.',
```
Can you extract this message to an external variable? Naming it `message` would it be destructured in.

j-f1 commented on src/rules/no-default-export.js:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    return {
+      ExportDefaultDeclaration: function (ast) {
```
Can you call this `node` to match the other rules?

j-f1 commented on src/rules/no-default-export.js:
```diff
@@ -0,0 +1,40 @@
+module.exports = {
+  meta: {
+    docs: {},
```
Can you put a `description` and category in here?

isiahmeadows replied:
I didn't include it because it seemed that none of the other rules appeared to use it ([example](https://github.com/benmosher/eslint-plugin-import/blob/master/src/rules/no-named-default.js)).

j-f1 commented on tests/src/rules/no-default-export.js:
```diff
@@ -0,0 +1,123 @@
+      }),
+
+    // ...SYNTAX_CASES,
+  ],
```
Can you add a test for something like `import defaultImport from 'module-from-npm'`? This should be legal, since we can’t control how other libraries are structured.

isiahmeadows replied:
I didn't add a test for it, since it only checks `export` and `export ... from`. I can add a test to clarify this, though.

isiahmeadows commented:
Ping, `eslint-plugin-import` people? (I feel this might have gotten lost in the weeds a little bit...)

ljharb commented:
@isiahmeadows if you rebase this (rather than using the "update branch" button), it should rerun the travis tests.

I'll give it a review, but I'm not a fan of even the existence of this rule, so I'm going to leave it to the other collaborators to review/merge.

ljharb commented on docs/rules/no-default-export.md:
```diff
@@ -0,0 +1,63 @@
+# no-default-export
```
i think maybe `forbid-default-export` would be a more typical name for a rule like this, and for the opposite of "prefer"

ljharb replied:
ping

isiahmeadows replied:
I was following the naming convention the other rules have - there's also `no-dynamic-require`, `no-named-as-default`, and `no-anonymous-default-export`. So IMHO it wouldn't make sense to use `forbid-${name}` when all the others use `no-`. Also, [ESLint recommends this convention](https://eslint.org/docs/developer-guide/working-with-rules#rule-naming-conventions), so I'm kind of hesitant to deviate from it

isiahmeadows replied:
BTW, I kind of missed these comments somehow; sorry about that. 😦

ljharb replied:
fair enough; this one isn't a blocker or anything.

j-f1 commented on CHANGELOG.md:
```diff
@@ -9,6 +9,7 @@ This change log adheres to standards from [Keep a CHANGELOG](http://keepachangel
 
 ## [2.8.0] - 2017-10-18
 ### Added
+- [`no-default-export`] rule ([#889], thanks [@isiahmeadows])
```
Don’t forget to add `[@isiahmeadows]: https://github.com/isiahmeadows` to the bottom of this file.

isiahmeadows replied:
Oh yeah...oops

ljharb commented on tests/src/rules/no-default-export.js:
```diff
@@ -0,0 +1,129 @@
+    test({
+      code: 'export { a, b } from "foo.js"',
+      parser: 'babel-eslint',
+      }),
```
all these `}),` need to be de-indented a level (it seems like this is already the way it is in the prefer-default-export tests, but let's fix it here anyways)

ljharb replied:
the indentation here is still off; the `}` should match the indentation level of the line that ends with `}`

ljharb commented on src/rules/no-default-export.js:
```diff
@@ -0,0 +1,33 @@
+      return {}
+    }
+
+    var message = 'Prefer named exports.'
```
this should be a `const`.

also, i wonder if this is an opportunity here to make the message a bit more explanatory?

ljharb replied:
ping

isiahmeadows replied:
Will do. I'll just note that the code base is a haphazard mix of `var`, `let`, and `const`, and ES6 features are used very inconsistently.

> also, i wonder if this is an opportunity here to make the message a bit more explanatory?

For the case of `export {foo as default}`, sure. For the rest, I'm not sure there is more that needs to be done IMHO. There's not really any extra context needed.

isiahmeadows commented:
@ljharb I rebased locally and resolved the issue.

FWIW, [this article](https://blog.neufund.org/why-we-have-banned-default-exports-and-you-should-do-the-same-d51fdc2cf2ad?gi=df38c1496fe2) just ran through JavaScript Weekly, and I feel it's highly relevant. I agree with most of what that article states, but it's purely an opinion, and although I happen to agree with most of it, I'm ***very*** against making it the default.

Also, not all languages have default imports/exports (in my experience, most with module systems built-in *don't*).

ljharb commented:
I've seen it - that article is full of "arguments" supporting named exports that apply identically to default exports; it doesn't make a convincing case at all.

Most languages also don't have a package manager that allows conflicting dependencies, but luckily JS isn't constrained by that :-)

ljharb commented on README.md:
```diff
@@ -78,7 +78,9 @@ This plugin intends to support linting of ES2015+ (ES6+) import/export syntax, a
 * Forbid anonymous values as default exports ([`no-anonymous-default-export`])
+* Forbid default exports ([`no-default-export`])
 * Prefer named exports to be grouped together in a single export declaration ([`group-exports`])
+* Forbid default exports ([`no-default-export`])
```
this rebase seems to have left this item here twice

ljharb commented on src/rules/no-default-export.js:
```diff
@@ -0,0 +1,33 @@
+
+      ExportNamedDeclaration: function (node) {
+        for (var i = 0; i < node.specifiers.length; i++) {
+          var specifier = node.specifiers[i]
```
this should use `node.specifiers.forEach` instead of a loop (which will allow both `var` instances to be removed)

isiahmeadows commented:
@ljharb All fixed (plus some other indentation oddities).

isiahmeadows commented:
@ljharb I added a new commit to address your other feedback (that I somehow missed), and I also fixed a bug in the process (with the proposed `export foo from "bar.js"`, which exports a named binding, not a default one).

isiahmeadows commented:
Ping, @ljharb/@benmosher?

ljharb commented:
@isiahmeadows I rebased this branch for you. It's still waiting on reviews from other collabs though.

isiahmeadows commented:
Ping @benmosher / @jfmengels?

isiahmeadows commented:
Ping? @benmosher / @jfmengels

sergeysova commented:
Any progress here?

sergeysova commented:
Add to readme?

isiahmeadows commented:
@sergeysova #1026
